### PR TITLE
example/st_things: [DO NOT MERGE] st_things demo on esp32_wrover board

### PR DESCRIPTION
**Pull request by zhouxinhe**
**DO NOT MERGE**, This PR is for demo only.

GPIO NUM 18/19 for reset/switch buttons
GPIO NUM 0/2/4 for internal RGB LEDs
GPIO NUM 21/22/23 for external RGB LEDs
Config CONFIG_DEMO_LED to enable LED control
- CONFIG_RESET_BUTTON=y && CONFIG_LIGHT_SWITCH_BUTTON=y  (using iotbus)
  then please config reset/sw button pin number: 18/19
- CONFIG_RESET_BUTTON && CONFIG_LIGHT_SWITCH_BUTTON is not set
  then a thread would be created for button scan, pin 18 for reset and pin 19 for switch

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [a5d998075f5d3f3117efbb67eb5db25fe522c4db] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [a5d998075f5d3f3117efbb67eb5db25fe522c4db] - Code Rule Check Result:
apps/examples/st_things/light/sample_light_LED.c:51: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:54: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
apps/examples/st_things/light/sample_light_LED.c:54: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:154: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:159: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:164: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:169: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:174: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:199: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_LED.c:204: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_LED.c:209: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_LED.c:230: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_LED.c:231: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_LED.c:234: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:237: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_LED.c:240: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_LED.c:243: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_LED.c:246: ERROR: [BRC_M_SMT] else should follow close brace '}'
apps/examples/st_things/light/sample_light_things.c:79: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_things.c:82: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_things.c:90: ERROR: [SPC_M_KWD] space required before the open brace '{'
apps/examples/st_things/light/sample_light_things.c:93: ERROR: [SPC_M_KWD] space required before the open brace '{'

**Comment by zhouxinhe** [comment]
This PR depends on #3239  and #3279.

File: apps/examples/st_things/light/sample_light_LED.c [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+int key_get_status(int key, int* status)
+{
+	int fd_key = -1;
+	int ret = ERROR;
+	char path[16] = { 0 };
+
+	if (key == KEY_RESET_IO_NUM) {
+		sprintf(path, gpio_path_format, KEY_RESET_IO_NUM);
+	}
```
Comment (NilabhShant) [inline apps/examples/st_things/light/sample_light_LED.c]: Please make it like: } else if {

File: apps/examples/st_things/light/sample_light_LED.c [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+static int led_set_status(int led, int status)
+{
+	char str[4] = { 0 };
+	int ret = ERROR;
+	int len = sprintf(str, "%d", status > 0);
+	if (led == LED_GREEN_IO_NUM) {
+		if (fd_led_green >= 0) {
+			ret = write(fd_led_green, str, len);
+		}
+	}
```
Comment (NilabhShant) [inline apps/examples/st_things/light/sample_light_LED.c]: same as above.

File: apps/examples/st_things/light/sample_light_LED.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+*
+****************************************************************************/
+
+#ifndef _SAMPLE_LIGHT_LED_H_
+#define _SAMPLE_LIGHT_LED_H_
+
+// Internal RGB LED config
+#define LED_RED_IO_NUM				0
+#define LED_GREEN_IO_NUM			2
+#define LED_BLUE_IO_NUM				4
```
Comment (NilabhShant) [inline apps/examples/st_things/light/sample_light_LED.h]: Please aline this macro

File: apps/examples/st_things/light/sample_light_LED.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+// External RGB LED config
+#define LED_EXT_RED_IO_NUM			21
+#define LED_EXT_GREEN_IO_NUM		22
```
Comment (NilabhShant) [inline apps/examples/st_things/light/sample_light_LED.h]: same as above

**Comment by seinfra** [comment]
Target : [2a4c1f83e2abea96b894604e125c54598c4f889d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [2a4c1f83e2abea96b894604e125c54598c4f889d] - Code Rule Check OK.

File: apps/examples/st_things/light/sample_light_LED.c (hunk 1 above)
Comment (zhouxinhe) [reply apps/examples/st_things/light/sample_light_LED.c]: I fixed all coding rule problems, thanks for your review!

File: apps/examples/st_things/Kconfig [hunk 5]
```diff
@@ -19,33 +18,45 @@ choice
...
 config EXAMPLES_ST_THINGS_MSG_HANDLING
 	bool "Message Handling Example"
+	select IOTBUS
 	select IOTBUS_GPIO
 	---help---
 		This is a Message handling Sample.
 
 endchoice
 
+config DEMO_LED
```
Comment (Hyunjun85) [inline apps/examples/st_things/Kconfig]: Why did you make new config for sample.
As you know there is EXAMPLE_ST_THINGS_LIGHT. 
It is not good to make new one. 

Could you please consider to use 'EXAMPLE_ST_XX'?

Comment (zhouxinhe) [reply apps/examples/st_things/Kconfig]: This PR is just used for demo with external LEDs easily.
I specify GPIO pins for LEDs and buttons in os/board/esp32_devkit/src/esp32_boot.c
and iotbus does not work on my side, so I also create new thread to scan button status...
Anyway, this is not a formal PR, I will close it later.
Thanks for your review :)

**Pull request closed without merging**